### PR TITLE
Dev sir 1542 erc20 provisioning

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Brokerage.Common/Domain/Processing/Context/OperationProcessingContext.cs [hunk 1]
```diff
@@ -1,49 +1,102 @@
...
-        public IReadOnlyCollection<TinyDeposit> TinyDeposits { get; }
+        public Blockchain Blockchain { get; }
+        public IReadOnlyCollection<Account> Accounts { get; }
+        public IReadOnlyCollection<BrokerAccount> BrokerAccounts { get; }
+        public IReadOnlyCollection<AccountDetails> AccountDetails { get; }
+        public IReadOnlyCollection<BrokerAccountDetails> BrokerAccountDetails { get; }
+        public IReadOnlyCollection<AccountBalances> AccountBalances => _accountBalances;
 
         public bool IsEmpty => Operation == null &&
                                !RegularDeposits.Any() &&
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Processing/Context/OperationProcessingContext.cs:77]: All deposits here?

File: src/Brokerage.Common/Domain/Processing/Context/TransactionProcessingContextBuilder.cs [hunk 2]
```diff
@@ -75,6 +78,9 @@ public sealed class TransactionProcessingContextBuilder
             var matchedBrokerAccountDetails = await brokerAccountDetailsRepository.GetAnyOf(matchedBrokerAccountDetailsIds);
             var matchedOperation = await operationsRepository.GetOrDefault(operationId);
 
+            if (operationId.HasValue && matchedOperation == null)
+                throw new InvalidOperationException($"Operation with id {operationId} has no operation");
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Processing/Context/TransactionProcessingContextBuilder.cs:82]: Sounds strange

File: src/Brokerage.Common/Domain/Processing/ProcessorsFactory.cs [hunk 3]
```diff
@@ -28,28 +28,35 @@ public ProcessorsFactory(IServiceProvider serviceProvider)
...
+                typeof(CompletedRegularDepositProcessor),
+                typeof(CompletedTokenDepositProvisioningProcessor),
+                typeof(CompletedTokenDepositProcessor),
                 typeof(CompletedWithdrawalProcessor)
             };
 
             _failedOperationProcessors = new List<Type>
             {
                 typeof(FailedDepositProcessor),
+                typeof(FailedProvisioningProcessor),
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Processing/ProcessorsFactory.cs:59]: It looks like we've forgotten about failed consolidation

File: src/Brokerage.Common/Migrations/20201103140643_TokenDeposits_AccountBalancesGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+namespace Brokerage.Common.Migrations
+{
+    public partial class TokenDeposits_AccountBalancesGenerator : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateSequence(
+                name: "id_generator_account_balances",
+                schema: "brokerage",
+                startValue: 107000000L);
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Migrations/20201103140643_TokenDeposits_AccountBalancesGenerator.cs:12]: Add new object here https://github.com/swisschain/Sirius/blob/master/development/readme.md

File: src/Brokerage.Common/Persistence/Accounts/AccountBalancesEntity.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Brokerage.Common.Persistence.Accounts
+{
+    [Table(name: Tables.AccountBalances)]
+    public class AccountBalancesEntity
+    {
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Accounts/AccountBalancesEntity.cs:11]: The external generator used for it actually

File: src/Brokerage.Common/Persistence/Accounts/AccountsBalancesRepositoryExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Threading.Tasks;
+using Brokerage.Common.Domain.Accounts;
+
+namespace Brokerage.Common.Persistence.Accounts
+{
+    public static class AccountsBalancesRepositoryExtensions
+    {
+        public static async Task<AccountBalances> GetAsync(this IAccountsBalancesRepository repo, AccountBalancesId id)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Accounts/AccountsBalancesRepositoryExtensions.cs:9]: This method exists in the repository as `Get`

File: src/Brokerage.Common/Persistence/Accounts/IBrokerAccountsBalancesRepository.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Brokerage.Common.Domain.Accounts;
+
+namespace Brokerage.Common.Persistence.Accounts
+{
+    public interface IAccountsBalancesRepository
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Accounts/IBrokerAccountsBalancesRepository.cs:7]: Name of the class doesn't match the name of the file

File: src/Brokerage.Common/Persistence/Accounts/AccountBalancesEntity.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Brokerage.Common.Persistence.Accounts
+{
+    [Table(name: Tables.AccountBalances)]
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Accounts/AccountBalancesEntity.cs:7]: This duplicates fluent EF model configuration

File: src/Brokerage.Common/Persistence/Accounts/AccountBalancesEntity.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    public class AccountBalancesEntity
+    {
+        [Key]
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Accounts/AccountBalancesEntity.cs:10]: This duplicates fluent EF model configuration

File: src/Brokerage.Common/Persistence/Deposits/DepositEntity.cs [hunk 10]
```diff
@@ -38,9 +38,12 @@ public DepositEntity()
         public IReadOnlyCollection<DepositSourceEntity> Sources { get; set; }
         public DateTimeOffset CreatedAt { get; set; }
         public DateTimeOffset UpdatedAt { get; set; }
-
         public decimal MinDepositForConsolidation { get; set; }
-
         public DepositTypeEnum DepositType { get; set; }
+        public long? ProvisioningOperationId { get; set; }
+
+        public IReadOnlyCollection<DepositFeeEntity> ConsolidationFees { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Deposits/DepositEntity.cs:45]: Estimated consolidation fees?

File: src/Brokerage.Common/Persistence/Deposits/DepositsRepository.cs [hunk 11]
```diff
@@ -60,6 +60,46 @@ public DepositsRepository(DatabaseContext dbContext, IDepositFactory depositFact
...
 
+        public async Task<IReadOnlyCollection<Deposit>> SearchForProvisioning(string blockchainId,
+            string transactionId,
+            long? provisioningOperationId)
+        {
+            var deposits = _dbContext
+                .Deposits
+                .AsQueryable();
+
+            if (blockchainId != null)
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Deposits/DepositsRepository.cs:71]: Do we have indexes for all these new conditions?

File: src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs [hunk 12]
```diff
@@ -85,12 +85,30 @@ public async Task<IReadOnlyCollection<MinDepositResidual>> GetForConsolidationDe
...
+            var minDepositResiduals = _dbContext.MinDepositResiduals
+                .AsQueryable()
+                .Where(x => x.ConsolidationDepositId != null &&
+                            consolidationDeposits.Contains(x.ConsolidationDepositId.Value));
 
-            await minDepositResiduals
-                .Where(x => x.ConsolidationDepositId != null && 
-                            consolidationDeposits.Contains(x.ConsolidationDepositId.Value))
-                .LoadAsync();
+            await minDepositResiduals.LoadAsync();
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs:93]: double round-trip to the DB here, as you've found recently

File: src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs [hunk 13]
```diff
@@ -85,12 +85,30 @@ public async Task<IReadOnlyCollection<MinDepositResidual>> GetForConsolidationDe
...
+        {
+            if (!provisioningDeposits.Any())
+                return Array.Empty<MinDepositResidual>();
+
+            var minDepositResiduals = _dbContext.MinDepositResiduals
+                .AsQueryable()
+                .Where(x => x.ProvisioningDepositId != null &&
+                            provisioningDeposits.Contains(x.ProvisioningDepositId.Value));
+
+            await minDepositResiduals.LoadAsync();
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs:111]: The same

File: src/Brokerage.Common/Domain/BrokerAccounts/BrokerAccountBalances.cs [hunk 14]
```diff
@@ -150,13 +150,50 @@ public void ConfirmBrokerWithDestinationTagPendingBalance(decimal amount)
 
         public void ConsolidateBalance(decimal receivedAmount, decimal fee)
         {
-            if (receivedAmount <= 0 || fee <= 0)
+            //if (receivedAmount <= 0 || fee <= 0)
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.Common/Domain/BrokerAccounts/BrokerAccountBalances.cs:153]: image
For convenience, I moved it to the upper level. Let's have a call about it

File: src/Brokerage.Common/Domain/BrokerAccounts/BrokerAccountBalances.cs [hunk 15]
```diff
@@ -150,13 +150,50 @@ public void ConfirmBrokerWithDestinationTagPendingBalance(decimal amount)
...
+            OwnedBalance -= fee;
+
+            UpdatedAt = DateTime.UtcNow;
+
+            GenerateEvent();
+        }
+
+        public void CompleteProvisioning(decimal provisionedBalance, decimal actualFee, decimal expectedFee)
+        {
+            if (provisionedBalance <= 0 || actualFee <= 0 || expectedFee <= 0)
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.Common/Domain/BrokerAccounts/BrokerAccountBalances.cs:168]: Eehhh, no -_-. The transaction was sent and confirmed here, we should receive non-zeros in each argument.

File: src/Brokerage.Common/Domain/BrokerAccounts/BrokerAccountBalances.cs [hunk 16]
```diff
@@ -150,13 +150,50 @@ public void ConfirmBrokerWithDestinationTagPendingBalance(decimal amount)
...
+            OwnedBalance -= actualFee;
+
+            UpdatedAt = DateTime.UtcNow;
+
+            GenerateEvent();
+        }
+
+        public void RevertProvisioning(decimal provisionedBalance, decimal actualFee, decimal expectedFee)
+        {
+            if (provisionedBalance <= 0 || actualFee <= 0 || expectedFee <= 0)
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.Common/Domain/BrokerAccounts/BrokerAccountBalances.cs:186]: Here, I guess you are right

File: src/Brokerage.Common/Domain/Deposits/Implementations/TokenDeposit.cs [hunk 17]
```diff
@@ -61,17 +69,69 @@ public void Fail(DepositError depositError)
...
             decimal residual,
-            Account account)
+            Account account,
+            Blockchain blockchain,
+            AccountBalances accountBalances,
+            BrokerAccountBalances brokerAccountBalances)
         {
             SwitchState(new[] { DepositState.Detected, }, DepositState.Confirmed);
 
+            var provisioningAmount = new Unit(
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Deposits/Implementations/TokenDeposit.cs:86]: right

File: src/Brokerage.Common/Domain/Deposits/MinDepositResidual.cs [hunk 18]
```diff
@@ -11,15 +11,19 @@ public class MinDepositResidual
...
             DateTimeOffset createdAt,
+            bool markForDeletion,
             uint version)
         {
             DepositId = depositId;
             Amount = amount;
             AccountDetailsId = accountDetailsId;
             AssetId = assetId;
             ConsolidationDepositId = consolidationDepositId;
+            ProvisioningDepositId = provisioningDepositId;
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Deposits/MinDepositResidual.cs:24]: Let me explain, each min deposit residual should belong to a deposit in the end in order to be included in (regular or token) deposit. This is the id of deposit for which action had to happen. We need to ids to process it on different stages

File: src/Brokerage.Common/Domain/Deposits/Processors/CompletedTokenDepositProvisioningProcessor.cs [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+                    brokerAccountDetails,
+                    accountDetails,
+                    _operationsFactory,
+                    margin,
+                    account,
+                    processingContext.Blockchain);
+
+                foreach (var residual in residuals)
+                {
+                    residual.AddConsolidationDeposit(deposit.Id);
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Deposits/Processors/CompletedTokenDepositProvisioningProcessor.cs:65]: Provisioning has been completed, so we start consolidation here. All residuals are included in the consolidation process

File: src/Brokerage.Common/Domain/Processing/Context/TransactionProcessingContextBuilder.cs (hunk 2 above)
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Processing/Context/TransactionProcessingContextBuilder.cs:82]: lol

File: src/Brokerage.Common/Domain/Processing/ProcessorsFactory.cs (hunk 3 above)
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Processing/ProcessorsFactory.cs:59]: Consolidation is included in failedDepositProcessor

File: src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs (hunk 12 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs:93]: You are wrong here there is no double round trip

File: src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs (hunk 13 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/Deposits/MinDepositResidualsRepository.cs:111]: nope

File: src/Brokerage.Common/Configuration/AppConfig.cs [hunk 20]
```diff
@@ -10,6 +10,13 @@ public class AppConfig
         public IndexerConfig Indexer { get; set; }
         public ExecutorConfig Executor { get; set; }
         public IReadOnlyDictionary<string, BlockchainProtocolConfig> BlockchainProtocols { get; set; }
-        public IReadOnlyDictionary<string, BlockchainConfig> Blockchains { get; set; }
+        public IReadOnlyCollection<AssetThresholdConfig> AssetThresholds { get; set; }
+    }
+
+    public class AssetThresholdConfig
```
[the remark this answers is not in the document]
Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Configuration/AppConfig.cs:16]: still actual

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**